xqueue_watcher/manager.py

```diff
@@ -133,29 +147,53 @@ def start(self):
             self.log.info('Starting %r', c)
             c.start()
 
-    def wait(self):
+    def wait(self, quit_if_empty=False):
         """
         Monitor clients.
         """
-        if not self.clients:
-            return
         signal.signal(signal.SIGTERM, self.shutdown)
+        config_disappeared = False
         while 1:
+            if not self.clients:
+                self.log.warning('No clients configured in %s', self.config_file)
+                if quit_if_empty:
+                    return
+            try:
+                time.sleep(self.manager_config['POLL_TIME'])
+            except KeyboardInterrupt:  # pragma: no cover
+                self.shutdown()
+            try:
+                # check for config file changes
+                if self.did_config_change():
+                    self.restart()
+                    config_disappeared = False
+            except FileNotFoundError:
+                if config_disappeared:
+                    self.log.error('Config file %s disappeared. Exiting', self.config_file)
+                    self.shutdown()
+                else:
+                    # in case the file was slow to move,
+                    # give one more try through the loop before restarting
+                    config_disappeared = True
+                    self.log.error('Config file %s disappeared. Retrying', self.config_file)
```

Comment:
Did this happen while you were developing? To what extent is xqueue watcher responsible for restarting itself (as opposed to whatever is invoking it doing the restart)?

Reply:
I was trying to anticipate what might happen if this used remote-config. Remote config (hermes) polls a URL and replaces a configuration file, then optionally will run a command. In the other IDAs, hermes is configured to run `supervisorctl restart whatever` after the file changes. I don't think it's necessary to run xqueue_watcher under supervisord, as it requires more configuration and moving parts, especially if it's running under docker (which can already handle the case of crashed processes). Instead, the process looks for changes and cleanly restarts itself. I didn't check to see if hermes does an atomic move of the file, or if it writes the new file in place. If it's not atomic, this `config_disappeared` stuff should handle the case.

If this were deployed using Docker, the host machine would store the `config.yml`, and it could run hermes to update the config automatically. I thought about having hermes run inside the docker container, but the hermes code isn't really in shape for easy distribution like that, and it would be tricky to configure it on for those who want it, versus those who just want to run with a static configuration.